# Processamento de linguagem natural e modelo de análise de sentimento.

A fim de que seja possível comparar os tópicos abaixo descritos com os itens explicitados pelo barema, encontra-se, nessa célula, um índice comparativo entre ambas partes. Ele foi feito para facilitar a correção dos professores e permitir a existência de uma autonomia na nomenclatura das células de organização. 
- Instalação -> tópico 1.0
- Testes isolados - >  tópico 3.8 
- Definição de funções -> tópico 3.1 a 3.7
- Testes das funções -> tópico 4.3
- Análise descritiva -> tópico 2.0
- Pré processamento -> tópico 3.0
- Pipeline -> tópico 4.3
- Modelo Bag of words -> tópico 4.0


## 1.0 Instalação

Nessa parte do notebook realiza-se a instalação das bibliotecas que serão utilizadas no decorrer do código, bem como a importação da base de dados que baseia todas as posteriores análises. 

### 1.1 Importação das bibliotecas

In [ ]:
# instalação de pacotes para remoção de acentos
!pip install unidecode
# Instalando biblioteca Python "emoji"
!pip install emoji
# Instalação de biblioteca para remoção de stop words
!pip install -U spacy

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#importação das bibliotecas:
import pandas as pd
import numpy as np
# bibliotecas utilizadas para tokenização
from nltk.tokenize import word_tokenize
from nltk.tokenize import sent_tokenize
#biblioteca para tokenização 
from keras.preprocessing.text import Tokenizer
# bibliotecas utilizadas para remoção de stop words
import nltk
nltk.download('punkt')
#importação das bibliotecas para remoção de Stop Words
import spacy
nlp = spacy.cli.download('pt_core_news_sm')
nlp = spacy.load('pt_core_news_sm')
# bibliotecas para remoção de acentos
from unidecode import unidecode
# bibliotecas utilizadas para o tratamento de abreviações
import re
# Instalando biblioteca Python "emoji"
import emoji
# importação de biblioteca para exportação do modelo
import pickle
# importação de biblioteca para separar treino e teste
from sklearn.model_selection import train_test_split
# importação do naive Bayes
from sklearn.naive_bayes import GaussianNB

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


✔ Download and installation successful
You can now load the package via spacy.load('pt_core_news_sm')


### 1.2 Criação da tabela 

In [ ]:
#Conexão com o drive
from google.colab import drive
drive.mount('/content/drive')
#importação da base de dados:
df = pd.read_excel('/content/drive/MyDrive/Modulo_6_grupo_1/2-base_10052023-G1.xlsx')
novas_colunas = {
    '"dataPublicada"': 'dataPublicada',
    '"autor"': 'autor',
    '"texto"': 'texto',
    '"sentimento"': 'targetSentimento',
    '"tipoInteracao"': 'tipoInteracao',
    '"anomalia"': 'anomalia',
    '"probabilidadeAnomalia"': 'probabilidadeAnomalia',
    '"linkPost"': 'linkPost',
    '"processado"': 'processado',
    '"contemHyperlink"': 'contemHyperlink'
}

# Renomear as colunas
df = df.rename(columns=novas_colunas)
# Remover linhas sem a coluna "texto"
df = df.dropna(subset=['texto'])
# Filtrar as linhas com base em diferentes condições
df = df[(df['anomalia'] != 1) & (df['autor'] != 'btgpactual') & (df['contemHyperlink'] != 1) & (df['probabilidadeAnomalia'] <= 50)]
# Redefinir o índice
df = df.reset_index(drop=True)

Mounted at /content/drive


##### Importação da tabela

In [ ]:
#Conexão com o drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#importação da base de dados:
df = pd.read_excel('/content/drive/MyDrive/Modulo_6_grupo_1/2-base_10052023-G1.xlsx')
df

##### Colunas renomeadas

In [ ]:
novasColunas = {'"dataPublicada"': 'dataPublicada', '"autor"': 'autor', '"texto"': 'texto', '"sentimento"':'sentimento', '"tipoInteracao"':'tipoInteracao', '"anomalia"':'anomalia', '"probabilidadeAnomalia"': 'probabilidadeAnomalia', '"linkPost"':'linkPost', '"processado"': 'processado', '"contemHyperlink"': 'contemHyperlink'}
df = df.rename(columns=novasColunas)
df


##### Remoção de colunas que não possuem textos

In [ ]:
#remoção de todas as linhas que não possuem a coluna texto:
df = df.dropna(subset=['texto'])
df

##### Remoção de linhas que possuem anomalias

In [ ]:
df = df[df['anomalia'] != 1]

##### Remoção de comentários do banco BTG

In [ ]:
df = df[df['autor'] != 'btgpactual']
df

##### Remoção de linhas que possuem hyperlink

In [ ]:
df = df[df['contemHyperlink'] != 1]
df

##### Remoção de linhas que possuem probabilidade de anomalia maior que 50%

In [ ]:
df = df[df['probabilidadeAnomalia'] <= 50]
df = df.reset_index(drop=True)
df

##### Balanceamento dos dados com undersampling

In [ ]:
# # Separa os elementos das classes POSITIVE, NEUTRAL e NEGATIVE em três tabelas
# df_classe1 = df.loc[df['sentimento'] == "POSITIVE"]
# df_classe2 = df.loc[df['sentimento'] == "NEUTRAL"]
# df_classe3 = df.loc[df['sentimento'] == "NEGATIVE"]

# # Mantém 988 elementos nas tabelas df_classe1 e df_classe2
# df_classe1 = df_classe1.sample(n=988, random_state=42)
# df_classe2 = df_classe2.sample(n=988, random_state=42)

# # Concatena as três tabelas em uma tabela df1
# df1 = pd.concat([df_classe1, df_classe2, df_classe3], ignore_index=True)

In [ ]:
# # Embaralha as linhas da tabela na variável df
# df = df1.sample(frac=1).reset_index(drop=True)

# # Exibe a tabela resultante
# df['sentimento']

0        NEUTRAL
1       POSITIVE
2        NEUTRAL
3        NEUTRAL
4       NEGATIVE
          ...   
3947    NEGATIVE
3948    NEGATIVE
3949    NEGATIVE
3950    NEGATIVE
3951     NEUTRAL
Name: sentimento, Length: 3952, dtype: object

## 2.0 Análise descritiva

A Análise Descritiva dos Dados é uma técnica importante para obter insights e compreender as características dos dados sem a utilização de modelos estatísticos complexos.

### 2.1 Colunas

Célula que permite a visualização de todas as colunas do dataframe:

In [ ]:
df.columns


1. **Coluna Id:** A coluna id apresenta o index para visualização da planilha e pode ser usada como chave primária do comentário. No que diz respeito à utilização no modelo, essa coluna não confere relevância para a sua construção, visto que sua utilidade está atrelada apenas ao fato de garantir que cada linha possua um identificador único. Portanto, *essa coluna não será utilizada.*

2. **Coluna dataPublicada:** A coluna data publicada refere-se a data de publicação do comentário. Para a construção do modelo **bag of words**, essa coluna não apresenta relevância, portanto, *não será utilizada*. Posteriormente, sua utilização pode se fazer necessária para a averiguação dos períodos das campanhas.

3. **Coluna autor:** A coluna autor é referente à conta do instagram que realizou o comentário na postagem. Essa coluna não será utilizada diretamente pelo modelo bag of words, mas será importante para o agrupamento de comentários referente à empresa BTG. 

4. **Coluna texto:** A coluna texto se refere ao texto presente no comentário realizado. Para a construção do modelo, essa é a coluna com maior relevância, visto que são justamente os conteúdos dos comentários que precisam ser analisados pelo modelo. 

5. **Coluna sentimento:** A coluna sentimento é, justamente, o target da classificação que precisamos fazer dos dados. Essa coluna será utilizada para o treinamento posterior do modelo, visto que apresenta o resultado esperado. Contudo, diante de uma análise manual das classificações, foi possível perceber que alguns comentários foram classificados de maneira errônea, dado que possuem um teor positivo mas foram classificados como negativos ou neutros.

6. **Coluna tipoInteracao:** A coluna tipo interação informa o tipo de interação a qual aquele comentário pertence, como, por exemplo, uma resposta ou como marcação. Essa coluna, inicialmente, não será utilizada para a construção do modelo bag of words.

7. **Coluna anomalia:**  Não possuímos informações suficientes para definir qual é o significado dessa coluna. Portanto, ainda não pode-se definir se ela será utilizada ou não no decorrer do desenvolvimento do projeto. Por ora, essa coluna não será utilizada.

8. **Coluna probabilidadeAnomalia:**  Não possuímos informações suficientes para definir qual é o significado dessa coluna. Portanto, ainda não pode-se definir se ela será utilizada ou não no decorrer do desenvolvimento do projeto. Por ora, essa coluna não será utilizada.

9. **Coluna linkPost:** Essa coluna possui o link referente a postagem da qual foram retirados os comentários. Todos os comentários referentes à mesma postagem possuem o link igual. Pensando na análise de sentimento, essa coluna não apresenta relevância, portanto, não será utilizada. 

10. **Coluna processado:** Não possuímos informações suficientes para definir qual é o significado dessa coluna. Portanto, ainda não pode-se definir se ela será utilizada ou não no decorrer do desenvolvimento do projeto. Por ora, essa coluna não será utilizada.

11. **Coluna contemHyperlink:** Não possuímos informações suficientes para definir qual é o significado dessa coluna. Portanto, ainda não pode-se definir se ela será utilizada ou não no decorrer do desenvolvimento do projeto. Por ora, essa coluna não será utilizada.



### 2.2 Funções

Abaixo estão listadas todas as funções presentes nesse notebook e o que elas fazem:
1. **remocaoAcentos:** função que remove todos os acentos presentes nos comentários da coluna texto.
2. **tratamentoMaiusculas:** função que trata todas as letras maiúsculas dos comentários presentes na coluna texto, passando-as para caixa baixa.
3. **conversaoEmojis:** função que converte todo emoji presente na coluna texto para seu significado em inglês.
4. **tratamentoAbreviacoes:** função que converte todas as abreviações para seu significado real, além de remover todas as pontuações e diferentes tipos de marcações comumente utilizadas em redes sociais.
5. **tokenizacaoPalavra:** função que tokeniza em palavras os comentários, separando cada uma delas do todo.
6. **tokenizacaoFrase:** função que tokeniza em frases os comentários, separando cada uma das frases, sejam elas do mesmo comentário ou de comentários diferentes.
7. **remocaoStopWords:** função que remove todas as stopwords presentes nos comentários da coluna texto.
8. **bagOfWords:** função para chamar o modelo de bag of words e salvar um vetor de recorrencia.
9. **pipelineBow:** função que define a pipeline que será utilizada no modelo, com todos os pré-processamentos realizados e o modelo BOW aplicado para gerar um dicionário e um vetor.
10. **contarPalavras:** função para contar quantas vezes cada palavra se repetiu no modelo bag of words.
11. **gerarNuvemPalavras:** função para gerar uma nuvem de palavras utilizando os resultados do modelo bag of words.

###2.3 Visualização gráfica de dados

#### 2.3.1 Valores nulos

Visualização da proporção entre o número de linhas que possuem algum tipo de valor nulo e as linhas que não os possuem. 

In [ ]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# visualizando todos os itens nulos
print(df.isnull().sum())

# criando um gráfico de barras para comparar os itens nulos e não nulos
sns.countplot(x=df.isnull().any(axis=1), hue=df.isnull().any(axis=1))
plt.title("Itens Nulos vs Não Nulos")
plt.xlabel("Itens")
plt.ylabel("Contagem")
plt.legend(labels=['Itens com valores não nulos', 'Itens com valores nulos'])
plt.show()

Visualização de ítens nulos em cada uma das colunas do dataframe:

In [ ]:
# contando os valores nulos em cada coluna
null_counts = df.isnull().sum()

# criando um gráfico de barras para mostrar a contagem de valores nulos em cada coluna
sns.barplot(x=null_counts.index, y=null_counts.values)
plt.xticks(rotation=90) # adicionando a rotação de 90 graus nas legendas

# adicionando as etiquetas de texto no topo de cada barra
for i, v in enumerate(null_counts.values):
    plt.text(i, v+10, str(v), ha='center', fontsize=12)

# configurando os rótulos dos eixos e o título do gráfico
plt.xlabel("Colunas")
plt.ylabel("Contagem de Valores Nulos")
plt.title("Contagem de Valores Nulos por Coluna")

# exibindo o gráfico
plt.show()

#### 2.2.2 Agrupamento por autores

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

# Agrupamento dos textos por autor, excluindo o "btg pactual"
textosPorAutor = df.query("autor != 'btgpactual'").groupby('autor')['texto'].count().reset_index()

# Ordenação dos autores por quantidade de textos
textosPorAutor = textosPorAutor.sort_values('texto', ascending=False).reset_index(drop=True)

# Seleção dos 10 autores com mais textos
top10Autores = textosPorAutor.head(10)

# Gráfico de barras com a quantidade de textos por autor
plt.bar(top10Autores['autor'], top10Autores['texto'])
plt.xticks(rotation=90)
plt.xlabel('Autor')
plt.ylabel('Quantidade de Textos')
plt.title('Top 10 Autores com Mais Textos (exceto btgpactual)')
plt.show()


### Análise dos sentimentos

In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

# Contar a quantidade de dados para cada combinação de marcação e tipo de interação
contagem = df.groupby(['tipoInteracao', 'sentimento']).size().unstack(fill_value=0)

# Criar o gráfico de barras agrupadas
contagem.plot(kind='bar', stacked=True)

# Configurar os rótulos dos eixos
plt.xlabel('Tipo de Interação')
plt.ylabel('Quantidade de Dados')

# Exibir o gráfico
plt.show()


In [ ]:
import pandas as pd
import matplotlib.pyplot as plt

# Contar a quantidade de dados para cada combinação de marcação e tipo de interação
contagem = df.groupby(['tipoInteracao', 'sentimento']).size().unstack(fill_value=0)

# Cores das colunas
cores = [ '#FF0000', '#C0C0C0', '#008000']

# Criar o gráfico de barras agrupadas com as cores personalizadas
contagem.plot(kind='bar', stacked=True, color=cores)

# Configurar os rótulos dos eixos
plt.xlabel('Tipo de Interação')
plt.ylabel('Quantidade de Dados')

# Exibir o gráfico
plt.show()


## 3.0 Pré processamento dos dados

O pré processamento dos dados é uma técnica que consiste em uma série de etapas que buscam preparar os dados para uma análise posterior, eliminando informações desnecessárias, corrigindo erros e padronizando o formato dos dados. 
Essa é uma etapa muito importante porque, muitas vezes, dados textuais são frequentemente não estruturados, o que significa que não seguem um formato padronizado ou organizado, dificultando a análise e extração de informações relevantes.

### 3.1 Remoção de acentos

Para garantir que um texto seja processado de forma consistente, é importante realizar a normalização dos seus componentes. Nesse sentido, a remoção de acentos é uma etapa fundamental, já que isso ajuda o algoritmo a compreender com mais precisão o que o texto está dizendo. Além disso, a ausência de acentos reduz o risco de que o algoritmo interprete de maneira diferente duas palavras idênticas, mas que tenham sido acentuadas de formas distintas ou incorretas. Em resumo, a remoção de acentos é uma prática essencial no pré-processamento de textos para aprimorar a qualidade e a eficácia da análise de dados em linguagem natural.

In [ ]:
#Código para remoção dos acentos
def remocaoAcentos(dataframe):
    dataframe = dataframe.applymap(lambda x: unidecode(str(x)))
    return dataframe

Teste isolado

In [ ]:
# Teste isolado da função remoção de acentos
remocaoAcentos(df)

### 3.2 Remoção letras maiusculas

O tratamento de letras maiúsculas é importante porque, muitas vezes, a diferença entre letras maiúsculas e minúsculas pode impactar negativamente a análise do algoritmo, tendo em vista que ele pode interpretar as palavras como diferentes em determinadas situações. Esse problema pode levar a resultados imprecisos, tendo em vista que, diante do objetivo de contar a frequência de palavras, essas diferenças garantem que as palavras sejam contadas de maneira distinta. 

In [ ]:
def tratamentoMaiusculas(dataframe):
    # Pré-processamento das letras maiúsculas
    dataframe = dataframe.applymap(lambda x: x.lower() if isinstance(x, str) else x)
    # Exibição dos dados pré-processados
    return dataframe

Teste isolado

In [ ]:
#teste isolado da função tratamento Maiúsculas
tratamentoMaiusculas(df)

### 3.3 Lematização

A lematização é uma técnica utilizada a fim de reduzir palavras flexionadas para sua forma básica. Ela representa as palavras através do infinitivo dos verbos e singular dos substantivos e adjetivos. Esse processo também leva em consideração o contexto inserido de cada palavra e resolve o problema da ambiguidade lexical, onde uma mesma palavra pode ter múltiplos significados.

In [ ]:
# Função de lematização
def textoLemmatizado(coluna):
    textosLemmatizados = []
    for texto in coluna:
        # Processa o texto com o modelo do Spacy
        doc = nlp(texto)
        # Lematiza cada token no documento
        lemmas = [token.lemma_ for token in doc]
        # Junta os lemmas em um único texto
        textoLemmatizado = ' '.join(lemmas)
        textosLemmatizados.append(textoLemmatizado)
    return textosLemmatizados

In [ ]:
#Teste isolado:
resultado = textoLemmatizado(df['texto'])
resultado

### 3.4 Tratamento de emojis

Emojis são muito utilizados por usuários de redes sociais, ajudando-os, assim, a expressar emoções e sentimentos. Portanto, para melhor entender o sentido de uma mensagem, o tratamento de emojis podem fornecer uma melhor precisão para a análise e uma classificação mais precisa dos sentimentos de uma mensagem.

In [ ]:
# Função para converter emojis em texto.
# A função recebe dois argumentos. O dataframe e a coluna a ser tratada, e retorna o dataframe.
def conversaoEmojis(coluna):
    coluna = coluna.apply(emoji.demojize)
    return coluna

Teste isolado

In [ ]:
# Teste isolado da função converter emoji em texto
conversaoEmojis(df['texto'])

### 3.5 Tratamento de abreviações e pontuações

O tratamento de abreviações é importante porque garante que o algoritmo de processamento de linguagem natural interprete corretamente o significado do texto. Ao substituir abreviações por suas formas completas, o texto se torna mais legível e compreensível para o algoritmo que realizará o processamento.

In [ ]:
   
def tratamentoAbreviacoes(coluna):    
    coluna = coluna.astype(str)
    coluna.str.lower()
# Dicionário com as abreviações e suas expansões
    abreviacoes = {
        r'\bvc\b': 'você',
        r'\btbm\b': 'também',
        r'\bpq\b': 'porque',
        r'\bmt\b': 'muito',
        r'\bmto\b':'muito',
        r'\bblz\b': 'beleza',
        r'\bjah\b': 'já',
        r'\bbj\b': 'beijo',
        r'\bflw\b': 'falou',
        r'\bvlw\b': 'valeu',
        r'\bkrl\b': 'caralho',
        r'\bqnd\b': 'quando',
        r'\bfds\b': 'fim de semana',
        r'\bqq\b': 'qualquer',
        r'\bmsg\b': 'mensagem',
        r'\bpra\b': 'para',
        r'\bp\b': 'para',
        r'\bpqp\b': 'puta que pariu',
        r'\bvlh\b': 'velho',
        r'\bnss\b': 'nossa',
        r'\bngm\b': 'ninguém',
        r'\bqm\b': 'quem',
        r'\bobs\b': 'observação',
        r'\bqt\b': 'quanto',
        r'\btbm ñ\b': 'também não',
        r'\bpf\b': 'por favor',
        r'\bpdc\b': 'pode crer',
        r'\btalkei\b': 'ok',
        r'\bd+\b': 'demais',
        r'\bdms\b': 'demais',
        r'\bsqn\b': 'só que não',
        r'\bmsm\b': 'mesmo',
        r'\bn/ao\b': 'não',
        r'\bn/\b': 'não',
        r'\bblz\b': 'beleza',
        r'\bq\b': 'que',
        r'\btd\b': 'tudo',
        r'\bpfvr\b': 'por favor',
        r'\bgnt\b': 'gente',
        r'\bto\b': 'estou',
        r'\bvcs\b': 'vocês',
        r'\bn/oa\b': 'não',
        r'\bsla\b': 'sei lá',
        r'\bflws\b': 'falou',
        r'\bbjos\b': 'beijos',
        r'\bkd\b': 'cadê',
        r'\blgl\b': 'legal',
        r'\bnum\b': 'número',
        r'\bjá q\b': 'já que',
        r'\bmano\b': 'irmão',
        r'\bpke\b': 'porque',
        r'\bpr\b': 'para',
        r'\btpm\b': 'tensão pré-menstrual',
        r'\btá\b': 'está',
        r'\bqmto\b': 'muito',
        r'\bñ\b': 'não',
        r'\bvamu\b': 'vamos',
        r'\btmj\b': 'estamos juntos',
        r'[0-9]': '',
        r'@\w+': '',
        r'[^\w\s]': '',
        r'\bwww\.[^\s]*': '',
        r'\bola\b': '',
        r'\n': '',
        r'\bbtg\b': '',
        r'\bpactual\b': ''
    }

    # Aplica a substituição de cada abreviação no dataframe
    for abreviacao, expansao in abreviacoes.items():
        coluna= coluna.apply(lambda x: re.sub(abreviacao, expansao, x, flags=re.IGNORECASE))
    
    return coluna

Teste isolado:

In [ ]:
#teste isolado
tratamentoAbreviacoes(df['texto'])

### 3.6 Tokenização

A tokenização é um processo de pré-processamento utilizado para dividir os valores de uma coluna em pedaços menores, como palavras ou frases. Cada pedaço, conhecido como token, recebe um valor específico para identificação. Esse processo torna o texto mais gerenciável e facilita a análise e os processamentos subsequentes dos dados.

Função para gerar um dicionario, com os tokens

In [ ]:
#definição de função para tokenização em palavras
def tokenizacaoPalavra(comentarios):
  tokenizer = Tokenizer()
  tokenizer.fit_on_texts(comentarios)
  return tokenizer

Tokenização por frase

In [ ]:
# definição de função para tokenização em frases
def tokenizacaoFrase(coluna):
  comentarios = coluna.values
  tokenFrase = []
  for comentarios in comentarios:
    tokenFrase.append(sent_tokenize(comentarios))
  return tokenFrase

Função para gerar tokens de palavras, referentes a cada frase

In [ ]:
def tokenizar(coluna):
    # Tokeniza as frases
    tokensPalavras = nltk.word_tokenize(coluna)
    return tokensPalavras

Testes isolados:

Geração de dicionario.

In [ ]:
tokenizacaoPalavra(df['texto'])

Tokenizando frases.

In [ ]:
tokenizacaoFrase(df['texto'])

Tokenizando as palavras de cada post individualmente

In [ ]:
resultado = df['texto'].apply(tokenizar)
resultado

### 3.7 Remoção de stopwords


A remoção dos Stop Words é um pré-processamento importante pois elimina palavras irrelevantes que podem prejudicar a precisão do modelo final, como artigos, preposições, conjunções, entre outros conectores. Esse tratamento ajuda na eficácia da classificação de texto  e na redução tanto do vocabulário quanto de ruídos (palavras que não têm um valor semântico significativo).

In [ ]:
def remocaoStopWords(coluna):
    listaSemStopWords = []
    for texto in coluna:
        # converte o texto em um documento do Spacy
        doc = nlp(texto)
        # filtra os tokens que não são stop words
        tokenSemStopwords = [token.text for token in doc if not token.is_stop]
        # adiciona a lista de frases sem stop words
        listaSemStopWords.append(' '.join(tokenSemStopwords))
    # substitui a coluna original pelos textos sem stop words
    coluna[:] = listaSemStopWords
    return coluna

Teste isolado

In [ ]:
remocaoStopWords(df['texto'])

### 3.8 Testes isolados

Teste isolado para função de remoção de acentos:

In [ ]:
remocaoAcentos(df)

Teste isolado para função de tratamento de letras maiúsculas:

In [ ]:
tratamentoMaiusculas(df)

teste isolado para função de conversão de emojis:

In [ ]:
conversaoEmojis(df['texto'])

Teste isolado para função de tratamento de abreviações:

In [ ]:
tratamentoAbreviacoes(df['texto'])

Teste isolado para função de gerar dicionario dos comentários:

In [ ]:
tokenizacaoPalavra(df['texto'])

Teste isolado da função de gerar tokens de cada comentario:

In [ ]:
resultado = df['texto'].apply(tokenizar)
resultado

Teste isolado para função que remove stopwords:

In [ ]:
remocaoStopWords(df['texto'])

### 3.9 Pipeline dos pré processamentos

Criação de pipiline para tratamento dos dados.

In [ ]:
def pipeline(dados):
  #cria um novo df a ser tratado
  dadosT =  pd.DataFrame({'dadosTratados': dados})
  # remove os acentos dos dados
  dadosT['dadosTratados'] = remocaoAcentos(dadosT) 
  # substituição das letras maiusculas dos dados
  dadosT['dadosTratados'] = tratamentoMaiusculas(dadosT)
  ## tratamento dos emojis
  #dadosT = conversaoEmojis(dadosT)
  #tratamento de abreviações 
  dadosT['dadosTratados'] = tratamentoAbreviacoes(dadosT['dadosTratados'])
  ##tratamento de stopwords
  dadosT['dadosTratados'] = remocaoStopWords(dadosT['dadosTratados'])
  ## lemmatização
  dadosT['dadosTratados'] = textoLemmatizado(dadosT['dadosTratados'])
  return dadosT['dadosTratados']

#### 3.9.1 Resultado dos pré processamentos.

Abaixo e póssivel ver o resultado dessa pipeline.

In [ ]:
dadosTratados = pipeline(df['texto'])
dadosTratados

0       alvarez    marsal estarar Conosco sportainmetr...
1       repost btgpactual With Make_repost    entendar...
2                                       minuto touro ouro
3                                                        
4                querer    banking proprio    administrar
                              ...                        
9482                           atendar telefone amor Deus
9483    voce sao    grande fiis mercado nao     seleci...
9484    erro financeiro eliminar    ano    parar osten...
9485    guadakan chegar    aula publico formato espeta...
9486       morning call nao aparecer spotify atualizar...
Name: dadosTratados, Length: 9487, dtype: object

Também foi criada a váriavel "dadosTratados", que pode ser utilizada posteriormente para visualizar o resultado de essa pipeline.

#### 3.9.2 Exportando pré processamento

In [ ]:
dfTratado = pd.DataFrame({'texto': dadosTratados, 'targetSentimento': df['targetSentimento']})

In [ ]:
dfTratado

texto targetSentimento
0     alvarez    marsal estarar Conosco sportainmetr...          NEUTRAL
1     repost btgpactual With Make_repost    entendar...          NEUTRAL
2                                     minuto touro ouro         POSITIVE
3                                                                NEUTRAL
4              querer    banking proprio    administrar         POSITIVE
...                                                 ...              ...
9482                         atendar telefone amor Deus         POSITIVE
9483  voce sao    grande fiis mercado nao     seleci...         POSITIVE
9484  erro financeiro eliminar    ano    parar osten...          NEUTRAL
9485  guadakan chegar    aula publico formato espeta...          NEUTRAL
9486     morning call nao aparecer spotify atualizar...         NEGATIVE

[9487 rows x 2 columns]

In [ ]:
dfTratado.to_csv('resultadoTratamento.csv', index=False)

##4.0  Bag of words

O modelo de bag of words (ou, em tradução literal, saco de palavras) é uma representação que transforma texto em vetores de tamanho fixo para contar quantas vezes as palavras aparecem nas frases em questão. Esse método é comumente chamado de **vetorização**.

### 4.1 Setup do modelo

In [ ]:
#importação de bibliotecas
#importação de keras, para execução do modelo e criação do dicionario
from keras.preprocessing.text import Tokenizer

#### 4.1.1 Definição de funções


função para criação dos vetores:


In [ ]:
#função para chamar o modelo de bag of words, e salvar um vetor de recorrencia.
def bagOfWords(tokens,comentarios):
  salvar = tokens.texts_to_matrix(comentarios, mode='count')
  return salvar

###4.2 Execução do modelo


##### 4.2.1 Teste do modelo de bag of words

Nesse momento é chamada a função bag of words para realização do teste da função.

In [ ]:
bagOfWords(tokenizacaoPalavra(df['texto']),df['texto'])

array([[0., 1., 3., ..., 0., 0., 0.],
       [0., 1., 2., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       ...,
       [0., 6., 2., ..., 0., 0., 0.],
       [0., 6., 7., ..., 1., 1., 1.],
       [0., 0., 1., ..., 0., 0., 0.]])

Como é possível visualizar pelo resultado, com o modelo, conseguimos obter como retorno uma matriz com o número de repetições de cada palavra presente no dicionário.

###4.3 Modelo Bag of Words

Abaixo é possível ver a função que define a pipeline que será utilizada no modelo, com todos os pré-processamentos realizados e o modelo BOW aplicado para gerar um dicionário e um vetor.

In [ ]:
def modeloBow(dados):
  #Recebe os dados a ser realizado o BOW
  dadosTratados = pipeline(dados)
  #cria um novo df a ser tratado
  dadosT =  pd.DataFrame({'dadosBow': dadosTratados})
  #tokenização/geração de dicionario
  dicionario = tokenizacaoPalavra(dadosT['dadosBow'])
  # chamando o modelo de bag of words, usando dados_t e o dicionario criado com a tokenização
  modelo = (bagOfWords(dicionario,dadosT['dadosBow']))
  return modelo,dicionario

Abaixo é possível ver o resultado da execução da função pipeline:

In [ ]:
#execução do modelo, e salvamento de variaveis
vetor,dicionario = (modeloBow(df['texto']))
dicionario = list(dicionario.word_index.keys())
print(dicionario)
print(vetor)

['btgpactual', 'nao', 'investimento', 'voce', 'banco', 'dia', 'mercado', 'ano', 'bom', 'financeiro', 'evento', 'ja', 'sao', 'brasil', 'fundo', 'so', 'cliente', 'hoje', 'empresa', 'contar', 'ser', 'tambem', 'r', 'time', 'acoes', 'ir', 'economia', 'taxa', 'ate', 'dinheiro', 'investir', 'ficar', 'pessoa', 'paulo', 'falar', 'ativo', 'semana', 'estar', 'bio', 'vida', 'fazer', 'negocio', 'oportunidade', 'investidor', 'ajudar', 'mercadofinanceiro', 'inflacao', 'aler', 'selic', 'de', 'futuro', 'compra', 'linkr', 'equipe', 'trabalho', 'ter', 'link', 'estao', 'parabem', 'fiis', 'politico', 'participar', 'cartao', 'incrivel', 'email', 'copom', 'importante', 'principal', 'la', 'sucesso', 'repost', 'passar', 'junto', 'resultado', 'mulher', 'mundo', 'conseguir', 'dar', 'abrir', 'credito', 'seguir', 'acontecer', 'cenario', 'brasileiro', 'alto', 'completo', 'realizar', 'especial', 'precisar', 'relatorio', 'parceiro', 'parceria', 'muito', 'conta', 'atendimento', 'mes', 'trader', 'pesquisa', 'gente', 'a

#### 4.3.1 Criação de dataframe com resultado final

Nesse dataframe encontra-se o resultado obtido após a execução da função pipeline.

In [ ]:
#adequação do vetor                                                                                                               #adequando o tamanho do vetor
vetorTabela = vetor[:, 1:]
#criando data frame
dfBagOfWords = pd.DataFrame(vetorTabela, columns=dicionario, index=dadosTratados)
dfBagOfWords

btgpactual  nao  \
dadosTratados                                                         
alvarez    marsal estarar Conosco sportainmetr ...         1.0  0.0   
repost btgpactual With Make_repost    entendar ...         1.0  1.0   
minuto touro ouro                                          0.0  0.0   
                                                           0.0  0.0   
querer    banking proprio    administrar                   0.0  0.0   
...                                                        ...  ...   
atendar telefone amor Deus                                 0.0  0.0   
voce sao    grande fiis mercado nao     selecio...         1.0  2.0   
erro financeiro eliminar    ano    parar ostent...         2.0  5.0   
guadakan chegar    aula publico formato espetac...         1.0  1.0   
   morning call nao aparecer spotify atualizar ...         0.0  1.0   

                                                    investimento  voce  banco  \
dadosTratados                                                                   
alvarez    marsal estarar Conosco sportainmetr ...           0.0   0.0    0.0   
repost btgpactual With Make_repost    entendar ...           0.0   0.0    0.0   
minuto touro ouro                                            0.0   0.0    0.0   
                                                             0.0   0.0    0.0   
querer    banking proprio    administrar                     0.0   0.0    0.0   
...                                                          ...   ...    ...   
atendar telefone amor Deus                                   0.0   0.0    0.0   
voce sao    grande fiis mercado nao     selecio...           0.0   1.0    0.0   
erro financeiro eliminar    ano    parar ostent...           1.0   0.0    0.0   
guadakan chegar    aula publico formato espetac...           0.0   1.0    0.0   
   morning call nao aparecer spotify atualizar ...           0.0   0.0    0.0   

                                                    dia  mercado  ano  bom  \
dadosTratados                                                                
alvarez    marsal estarar Conosco sportainmetr ...  0.0      0.0  0.0  0.0   
repost btgpactual With Make_repost    entendar ...  0.0      0.0  0.0  0.0   
minuto touro ouro                                   0.0      0.0  0.0  0.0   
                                                    0.0      0.0  0.0  0.0   
querer    banking proprio    administrar            0.0      0.0  0.0  0.0   
...                                                 ...      ...  ...  ...   
atendar telefone amor Deus                          0.0      0.0  0.0  0.0   
voce sao    grande fiis mercado nao     selecio...  0.0      1.0  0.0  0.0   
erro financeiro eliminar    ano    parar ostent...  0.0      0.0  2.0  0.0   
guadakan chegar    aula publico formato espetac...  1.0      0.0  1.0  0.0   
   morning call nao aparecer spotify atualizar ...  3.0      0.0  0.0  0.0   

                                                    financeiro  ...  anotem  \
dadosTratados                                                   ...           
alvarez    marsal estarar Conosco sportainmetr ...         0.0  ...     0.0   
repost btgpactual With Make_repost    entendar ...         0.0  ...     0.0   
minuto touro ouro                                          0.0  ...     0.0   
                                                           0.0  ...     0.0   
querer    banking proprio    administrar                   0.0  ...     0.0   
...                                                        ...  ...     ...   
atendar telefone amor Deus                                 0.0  ...     0.0   
voce sao    grande fiis mercado nao     selecio...         0.0  ...     0.0   
erro financeiro eliminar    ano    parar ostent...         2.0  ...     0.0   
guadakan chegar    aula publico formato espetac...         0.0  ...     1.0   
   morning call nao aparecer spotify atualizar ...         0.0  ...     0.0   

                               

#### 4.3.2 Exportação do resultado

Abaixo, é possível ver a geração de um arquivo csv, com o resultado do modelo:

In [ ]:
#Criação de csv com resultado do modelo
dfBagOfWords.to_csv('resultadoBOW.csv', index=False)

KeyboardInterrupt: ignored

### 4.4 Representação visual

##### 4.4.1 Tabela de frequência

Definição de função para contar quantas vezes cada palavra se repetiu.

In [ ]:
#código para gerar a matriz de repetição das palavras
def contarPalavras(dicionario, matriz):
    dicionario = [''] + dicionario
    contagemPalavras = [0] * len(dicionario)
    for i, palavra in enumerate(dicionario):
        for j in range(len(matriz)):
            contagemPalavras[i] += matriz[j][i]
    contagemDf = pd.DataFrame({'palavra': dicionario, 'repetição': contagemPalavras})
    return contagemDf.iloc[1:]


Chamando a função para "contarPalavras", para gerar uma tabela

In [ ]:
tabela = contarPalavras(dicionario, vetor)

Abaixo é possível visualizar quantas vezes cada palavra estava presente nos comentários fornecidos.

In [ ]:
tabela.head(10)

##### 4.4.2 Núvem de palavras

A nuvem de palavras foi utilizada como representação visual das palavras mais relevantes dos comentários analisados, nela o tamanho de cada palavra é proporcional à frequência em que ele aparece. Ela contribui com modelo de bow ao destacar os termos-chave e a resumir rapidamente o conteúdo mais expressivo dos comentários.

##### 4.4.2.1 setup

In [ ]:
#bibliotecas utilizadas
from wordcloud import WordCloud
import matplotlib.pyplot as plt

##### 4.4.2.2 Definição da função

Definindo função que ira gerar a nuvem de palavras

In [ ]:
#definição da função
def gerarNuvemPalavras(matrizRec, vocabulario):
    # Cria um dicionário com as palavras do vocabulário e suas frequências na matriz de recorrência
    freqPalavras = dict(zip(vocabulario, matrizRec.sum(axis=0)))
    
    # Cria uma instância da classe WordCloud com as configurações desejadas
    nuvemPalavras = WordCloud(width=8000, height=4000, background_color='white', colormap='viridis')
    
    # Gera a nuvem de palavras a partir do dicionário de frequências
    nuvemPalavras.generate_from_frequencies(freqPalavras)
    
    # Exibe a nuvem de palavras
    plt.figure(figsize=(12,6))
    plt.imshow(nuvemPalavras, interpolation='bilinear')
    plt.axis('off')
    plt.show()

##### 4.4.2.4 Visualização

Abaixo e possivel visualizar a nuvem de palavras:

In [ ]:
gerarNuvemPalavras(vetor, dicionario)

## 5.0 Word2vec pré - treinado

### 5.1 Setup do modelo

In [ ]:
#setup
import gensim
from scipy.spatial.distance import cosine
from gensim.models import KeyedVectors

In [ ]:
#Carregando o modelo
modelo_cbow = KeyedVectors.load_word2vec_format("/content/drive/MyDrive/Modulo_6_grupo_1/cbow_s50.txt")

### 5.2 definição de funções

In [ ]:
def word2Vec(coluna):
# lista que armazena os vetores resultantes de todas as frases
  x = []
  for i in range(0,len(coluna)):
# lista que armazena os vetores das palavras de cada frase
    vetor = []
    for h in range(0,len(coluna[i])):
      vetor.append(modelo_cbow[h])
# soma dos vetores de todas as palavras 
    x.append(list(map(sum, zip(*vetor))))
    vetor=[]
  print(x)
  return x

### 5.3 teste da funções

In [ ]:
# chama a função Word2Vec para testá-la
testeword2Vec = word2Vec(df['texto'].apply(tokenizar))
#criação de um Dataframe para visualizar o resultado
dfTesteword2Vec = pd.DataFrame(testeword2Vec)
dfTesteword2Vec

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



0          1         2         3         4         5         6   \
0    -0.514310  -7.478573 -1.852811 -7.445124 -0.900545 -0.450863 -0.844640   
1    -0.371537  -8.584228 -2.384340 -7.285557 -0.582353 -0.765648 -1.334225   
2     0.106028   0.121924  0.139373 -0.918912  0.004461 -0.216214 -0.204989   
3     0.047435  -0.007625  0.173723 -0.445078 -0.058977 -0.182599 -0.079506   
4     1.075528  -0.879334  0.146983 -2.114522  0.822631 -0.225273 -0.192567   
...        ...        ...       ...       ...       ...       ...       ...   
9482  0.572639  -0.125299 -0.170976 -1.169875  0.079236  0.001395 -0.204402   
9483 -0.175089  -7.093212 -1.770549 -7.163111 -1.447110 -0.746501 -0.146332   
9484 -1.227343  -8.025073 -5.392955 -6.965629  4.477837 -1.817244 -3.185648   
9485  1.730446 -10.583965 -5.965667 -7.616147  6.894613 -2.380157 -3.365684   
9486  1.650954  -2.281441 -0.887941 -3.605431  0.846856 -0.112471 -1.261535   

            7         8         9   ...        40        41        42  \
0     0.129873 -6.139715 -6.840437  ... -0.747759 -7.420995  1.201510   
1     0.468803 -5.871632 -7.197826  ... -0.219864 -7.189882  0.198363   
2    -0.420545 -0.648111 -0.363125  ... -0.179132 -0.510152  0.635353   
3    -0.220796 -0.265243 -0.346969  ... -0.012755 -0.238091  0.309302   
4    -1.433440 -2.440478 -1.470971  ...  0.587403 -2.146278  0.094727   
...        ...       ...       ...  ...       ...       ...       ...   
9482 -0.515468 -1.489781 -0.363050  ...  0.099337 -1.236094  0.378341   
9483  0.185476 -5.884923 -6.602993  ... -0.692300 -7.748579  1.333975   
9484  2.470718 -5.347249 -9.338262  ... -0.635413 -5.189282 -2.728264   
9485  3.469364 -1.747959 -7.804751  ... -0.918631 -0.695320 -7.331540   
9486 -0.753928 -4.107752 -3.310983  ...  0.856523 -3.308242 -1.157657   

            43         44        45        46        47         48        49  
0    -0.225418   2.372435  6.523726  1.322155  0.077827  -5.811534  6.145944  
1    -0.475961   3.055253  6.169072  0.200981  0.169087  -7.673683  6.896269  
2    -0.319989  -0.392927  0.184347 -0.225403  0.008890   0.201078  0.354131  
3    -0.140818  -0.078406 -0.180135  0.013944 -0.043409   0.288535  0.232324  
4    -0.327286   0.427182  0.367773 -0.884387  0.805210   0.683840  1.651200  
...        ...        ...       ...       ...       ...        ...       ...  
9482 -0.582291  -0.087733  0.685860 -0.503743  0.184030   0.816723  0.928673  
9483  0.273490   2.190751  6.459354  0.951604  0.148626  -5.948199  6.262340  
9484  0.192428   6.617338  8.040454 -2.140488 -1.201420 -11.837457  2.845104  
9485  0.490277  11.974003  9.637973 -0.909810 -4.649925 -17.022641  0.733577  
9486 -1.466912   0.360162  2.208855 -0.780562 -0.657792   0.071132  2.960655  

[9487 rows x 50 columns]

### 5.4 Teste isolado

Esses teste tem como objetivo comprovar que os vetores estão sendo corretamente somados

O codigo abaixo ainda não soma os vetores de cada palvra da frase

In [ ]:
#teste que converte palavras em vetores e retorna uma lista de vetores resultantes
def word2Vecteste(coluna):
#armazena os vetores resultantes
  x = []
  for i in range(0,len(coluna)):
    vetor = []
#armazena todos os vetores de todas as palavras
    for h in range(0,len(coluna[i])):
      vetor.append(modelo_cbow[h])
    x.append(vetor)
    vetor=[]
  print(x)
  return x

Abaixo esta sendo executado o teste da função

In [ ]:
# gera os vetores das 10 primeiras frases
testeSoma = word2Vecteste(df['texto'].apply(tokenizar).head(10))
#criação de um dataframe para retornar o valor do 'testeSoma' que contém a lista de vetores resultantes
dftesteSoma = pd.DataFrame(testeSoma)
testeSoma

#### Comprovação

Abaixo e possivel ver a 3 coluna do df['texto'], que sera utilizada como exemplo para esse teste.

In [ ]:
df['texto'][3]

'@ricktolledo Sim'

Abaixo e possivel ver o vetor de cada palavra da coluna 3 do df['texto'], que sera utilizada como exemplo para esse teste.

In [ ]:
testeSoma[3]

[array([-0.01    , -0.000949, -0.000157,  0.007802, -0.004775,  0.00916 ,
         0.000275,  0.003503,  0.002693, -0.000599,  0.002116,  0.007509,
         0.006414,  0.003036,  0.003444,  0.005029, -0.005577, -0.007318,
         0.001901,  0.007887, -0.00672 ,  0.008236, -0.000243, -0.009482,
         0.002523, -0.004002, -0.009337, -0.004473,  0.009624,  0.006939,
        -0.00016 ,  0.002124,  0.00215 , -0.003164, -0.003439, -0.009075,
        -0.009343, -0.00982 ,  0.009702, -0.009068, -0.006134, -0.008558,
         0.006518, -0.006201,  0.003556,  0.004342,  0.003762, -0.002697,
        -0.006839,  0.001996], dtype=float32),
 array([ 0.094044,  0.195893, -0.155718, -0.449506, -0.160736,  0.000748,
        -0.138087, -0.004585, -0.121695, -0.087401,  0.018703,  0.081959,
         0.106759, -0.016001, -0.009896, -0.162624,  0.014549,  0.010382,
         0.208517, -0.02957 ,  0.180953,  0.132365,  0.019936, -0.040737,
         0.019822,  0.131696, -0.08243 , -0.09606 ,  0.158126, -0

Como e possivel observar a primeira linha de cada correspondem ao valores -0.01, 0.094044, -0.036609

Abaixo e possivel visualizar o resultado da soma vetorial, no caso 0.047435

In [ ]:
dfTesteword2Vec[0][3]

0.04743499867618084

In [ ]:
testeSomaExemplo = - 0.01 + 0.094044 - 0.036609
testeSomaExemplo

0.047435000000000005

Como é possível observar, utilizando esse exemplo da coluna 3, a soma dos vetores das 3 palavras e 0.047435, assim como é possível observar no resultado da função original. Comprovando assim que a soma vetorial foi executada de forma correta.

### 5.5 Modelo Word2vec pré - treinado

#### 5.5.1 definição

Abaixo e definido modelo word2vec, funcionando em conjunto com a pipeline.

In [ ]:
def modeloWord2Vec(coluna):
  #Recebe os dados a ser realizado o BOW
  frases = pipeline(coluna)
  #Cria os tokens a serem utilizados pelo modelo
  frases = frases.apply(tokenizar)
  #aplica a função de word2vec
  vetores = word2Vec(frases)
  #retorna o reusltado
  return vetores

#### 5.5.2 teste

Abaixo e possivel vizualisar o modelo word2vec, funcionando em conjunto com a pipeline

In [ ]:
resultadoModeloWord2Vec= modeloWord2Vec(df['texto'])

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



####  5.5.3 Criando dataframe 

In [ ]:
dfWord2Vec = pd.DataFrame(resultadoModeloWord2Vec, index=dadosTratados)
dfWord2Vec

0         1   \
dadosTratados                                                            
alvarez    marsal estarar Conosco sportainmetr ...  1.383377 -2.654717   
repost btgpactual With Make_repost    entendar ...  1.283763 -3.410798   
minuto touro ouro                                   0.047435 -0.007625   
                                                         NaN       NaN   
querer    banking proprio    administrar            0.106028  0.121924   
...                                                      ...       ...   
atendar telefone amor Deus                          0.106028  0.121924   
voce sao    grande fiis mercado nao     selecio...  1.815789 -2.876974   
erro financeiro eliminar    ano    parar ostent...  0.836931 -4.082651   
guadakan chegar    aula publico formato espetac...  0.068602 -8.919511   
   morning call nao aparecer spotify atualizar ...  0.948231 -0.942366   

                                                          2         3   \
dadosTratados                                                            
alvarez    marsal estarar Conosco sportainmetr ... -1.405555 -3.668075   
repost btgpactual With Make_repost    entendar ... -2.465253 -3.282195   
minuto touro ouro                                   0.173723 -0.445078   
                                                         NaN       NaN   
querer    banking proprio    administrar            0.139373 -0.918912   
...                                                      ...       ...   
atendar telefone amor Deus                          0.139373 -0.918912   
voce sao    grande fiis mercado nao     selecio... -1.558797 -3.573577   
erro financeiro eliminar    ano    parar ostent... -2.630611 -4.756009   
guadakan chegar    aula publico formato espetac... -2.793904 -7.389237   
   morning call nao aparecer spotify atualizar ... -0.010881 -2.126990   

                                                          4         5   \
dadosTratados                                                            
alvarez    marsal estarar Conosco sportainmetr ...  0.229193  0.304061   
repost btgpactual With Make_repost    entendar ...  0.012297 -0.849639   
minuto touro ouro                                  -0.058977 -0.182599   
                                                         NaN       NaN   
querer    banking proprio    administrar            0.004461 -0.216214   
...                                                      ...       ...   
atendar telefone amor Deus                          0.004461 -0.216214   
voce sao    grande fiis mercado nao     selecio...  1.132767 -0.111063   
erro financeiro eliminar    ano    parar ostent... -0.629420 -0.144831   
guadakan chegar    aula publico formato espetac... -0.428109 -0.655916   
   morning call nao aparecer spotify atualizar ...  0.949057 -0.365847   

                                                          6         7   \
dadosTratados                                                            
alvarez    marsal estarar Conosco sportainmetr ... -1.807550 -0.759527   
repost btgpactual With Make_repost    entendar ... -1.143002 -0.586896   
minuto touro ouro                                  -0.079506 -0.220796   
                                                         NaN       NaN   
querer    banking proprio    administrar           -0.204989 -0.420545   
...                                                      ...       ...   
atendar telefone amor Deus                         -0.204989 -0.420545   
voce sao    grande fiis mercado nao     selecio... -1.510444 -0.775099   
erro financeiro eliminar    ano    parar ostent... -1.697202 -0.302105   
guadakan chegar    aula publico formato espetac... -1.413017  0.700860   
   morning call nao aparecer spotify atualizar ... -0.170032 -1.474218   

                                                          8         9   ...  \
dadosTratados                                                           ...   
alvarez    marsal estarar Conosco sportainmetr .

#### 5.5.4 exportando resultado

In [ ]:
dfWord2Vec.to_csv('resultadoWord2vec.csv', index=False)

## 6.0 Modelos

### 6.1 Naive Bayes com BagOfWords

#### Guto

#### 6.1.1 Naive bayes sem balanceamento

##### 6.1.1.1 Setup

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score
from imblearn.over_sampling import RandomOverSampler
from sklearn.utils import resample
from imblearn.over_sampling import SMOTE

##### 6.1.1.2 Definição de bases

In [ ]:
# Exemplo de uso da função
df_rotulos = df['targetSentimento'].copy()
df_bag_of_words = dfBagOfWords

In [ ]:
# Redefinir os índices dos DataFrames
df_rotulos = df_rotulos.reset_index(drop=True)
df_bag_of_words = df_bag_of_words.reset_index(drop=True)

# Juntar os DataFrames ao longo do eixo das colunas
df_unido = pd.concat([df_rotulos, df_bag_of_words], axis=1)
df_unido

targetSentimento  btgpactual  nao  investimento  voce  banco  dia  \
0             NEUTRAL         1.0  0.0           0.0   0.0    0.0  0.0   
1             NEUTRAL         1.0  1.0           0.0   0.0    0.0  0.0   
2            POSITIVE         0.0  0.0           0.0   0.0    0.0  0.0   
3             NEUTRAL         0.0  0.0           0.0   0.0    0.0  0.0   
4            POSITIVE         0.0  0.0           0.0   0.0    0.0  0.0   
...               ...         ...  ...           ...   ...    ...  ...   
9482         POSITIVE         0.0  0.0           0.0   0.0    0.0  0.0   
9483         POSITIVE         1.0  2.0           0.0   1.0    0.0  0.0   
9484          NEUTRAL         2.0  5.0           1.0   0.0    0.0  0.0   
9485          NEUTRAL         1.0  1.0           0.0   1.0    0.0  1.0   
9486         NEGATIVE         0.0  1.0           0.0   0.0    0.0  3.0   

      mercado  ano  bom  ...  anotem  convencoes  sessoes  lembramo  \
0         0.0  0.0  0.0  ...     0.0         0.0      0.0       0.0   
1         0.0  0.0  0.0  ...     0.0         0.0      0.0       0.0   
2         0.0  0.0  0.0  ...     0.0         0.0      0.0       0.0   
3         0.0  0.0  0.0  ...     0.0         0.0      0.0       0.0   
4         0.0  0.0  0.0  ...     0.0         0.0      0.0       0.0   
...       ...  ...  ...  ...     ...         ...      ...       ...   
9482      0.0  0.0  0.0  ...     0.0         0.0      0.0       0.0   
9483      1.0  0.0  0.0  ...     0.0         0.0      0.0       0.0   
9484      0.0  2.0  0.0  ...     0.0         0.0      0.0       0.0   
9485      0.0  1.0  0.0  ...     1.0         1.0      1.0       1.0   
9486      0.0  0.0  0.0  ...     0.0         0.0      0.0       0.0   

      coreografo  chico  neller  prefeituradecorumbar  cpanufmsoficial  \
0            0.0    0.0     0.0                   0.0              0.0   
1            0.0    0.0     0.0                   0.0              0.0   
2            0.0    0.0     0.0                   0.0              0.0   
3            0.0    0.0     0.0                   0.0              0.0   
4            0.0    0.0     0.0                   0.0              0.0   
...          ...    ...     ...                   ...              ...   
9482         0.0    0.0     0.0                   0.0              0.0   
9483         0.0    0.0     0.0                   0.0              0.0   
9484         0.0    0.0     0.0                   0.0              0.0   
9485         1.0    1.0     1.0                   1.0              1.0   
9486         0.0    0.0     0.0                   0.0              0.0   

      sebraems  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  
...        ...  
9482       0.0  
9483       0.0  
9484       0.0  
9485       1.0  
9486       0.0  

[9487 rows x 21298 columns]

##### 6.1.1.3 Criação da função 

In [ ]:
def train_naive_bayes_model(dataframe):
    # Passo 3: Divisão dos dados em features (X) e rótulos (y)
    X = dataframe.drop('targetSentimento', axis=1)  # Considerando que 'rótulos' é a coluna com os rótulos de classificação
    y = dataframe['targetSentimento']

    # Passo 4: Divisão dos dados em conjuntos de treinamento e teste
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Passo 5: Criação e treinamento do modelo Naive Bayes
    model = MultinomialNB()
    model.fit(X_train, y_train)

    # Passo 6: Fazer previsões nos dados de teste
    y_pred = model.predict(X_test)

    # Passo 7: Avaliar o modelo
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Acurácia do modelo: {accuracy}")

##### 6.1.1.4 Teste da função


In [ ]:
train_naive_bayes_model(df_unido)

Acurácia do modelo: 0.7118018967334035


### 6.2 Naive Bayes com Word2Vec

##### Izabella

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score
from imblearn.over_sampling import SMOTE

In [ ]:
# Exemplo de uso da função
df_rotulos = df['targetSentimento'].copy()
df_word_2_vec = dfWord2Vec

In [ ]:
# Redefinir os índices dos DataFrames
df_rotulos = df_rotulos.reset_index(drop=True)
df_word_2_vec = df_word_2_vec.reset_index(drop=True)
# Juntar os DataFrames ao longo do eixo das colunas
df_unido_word = pd.concat([df_rotulos, df_word_2_vec], axis=1)
df_unido_word

targetSentimento         0         1         2         3         4  \
0             NEUTRAL  1.383377 -2.654717 -1.405555 -3.668075  0.229193   
1             NEUTRAL  1.283763 -3.410798 -2.465253 -3.282195  0.012297   
2            POSITIVE  0.047435 -0.007625  0.173723 -0.445078 -0.058977   
3             NEUTRAL       NaN       NaN       NaN       NaN       NaN   
4            POSITIVE  0.106028  0.121924  0.139373 -0.918912  0.004461   
...               ...       ...       ...       ...       ...       ...   
9482         POSITIVE  0.106028  0.121924  0.139373 -0.918912  0.004461   
9483         POSITIVE  1.815789 -2.876974 -1.558797 -3.573577  1.132767   
9484          NEUTRAL  0.836931 -4.082651 -2.630611 -4.756009 -0.629420   
9485          NEUTRAL  0.068602 -8.919511 -2.793904 -7.389237 -0.428109   
9486         NEGATIVE  0.948231 -0.942366 -0.010881 -2.126990  0.949057   

             5         6         7         8  ...        40        41  \
0     0.304061 -1.807550 -0.759527 -4.674443  ...  0.220864 -3.715628   
1    -0.849639 -1.143002 -0.586896 -4.501272  ...  0.604381 -5.294922   
2    -0.182599 -0.079506 -0.220796 -0.265243  ... -0.012755 -0.238091   
3          NaN       NaN       NaN       NaN  ...       NaN       NaN   
4    -0.216214 -0.204989 -0.420545 -0.648111  ... -0.179132 -0.510152   
...        ...       ...       ...       ...  ...       ...       ...   
9482 -0.216214 -0.204989 -0.420545 -0.648111  ... -0.179132 -0.510152   
9483 -0.111063 -1.510444 -0.775099 -4.272242  ...  0.356593 -4.013429   
9484 -0.144831 -1.697202 -0.302105 -5.067372  ... -0.118279 -6.384274   
9485 -0.655916 -1.413017  0.700860 -5.643440  ... -0.731742 -7.833986   
9486 -0.365847 -0.170032 -1.474218 -2.592782  ...  0.838269 -2.599074   

            42        43        44        45        46        47        48  \
0    -0.508049 -1.681060  0.322339  3.070996 -1.145615 -0.220174 -1.008628   
1    -0.893327 -0.872390  1.264841  2.869610 -0.396956 -0.465566 -2.063515   
2     0.309302 -0.140818 -0.078406 -0.180135  0.013944 -0.043409  0.288535   
3          NaN       NaN       NaN       NaN       NaN       NaN       NaN   
4     0.635353 -0.319989 -0.392927  0.184347 -0.225403  0.008890  0.201078   
...        ...       ...       ...       ...       ...       ...       ...   
9482  0.635353 -0.319989 -0.392927  0.184347 -0.225403  0.008890  0.201078   
9483 -0.626634 -1.653492  0.486769  2.573709  0.071672 -0.287472 -0.189655   
9484 -0.833756  0.019192  1.295962  3.993481  0.281847  0.516516 -3.820503   
9485 -0.155092 -0.787291  3.670361  7.178771 -0.241885 -0.687680 -7.578391   
9486  0.021309 -0.552362  0.259679  0.921726 -0.890135  1.305274  0.611135   

            49  
0     4.302267  
1     4.795250  
2     0.232324  
3          NaN  
4     0.354131  
...        ...  
9482  0.354131  
9483  3.710586  
9484  6.278456  
9485  5.693573  
9486  1.765292  

[9487 rows x 51 columns]

In [ ]:
def train_naive_bayes_model_word(dataframe):
    # Passo 3: Divisão dos dados em features (X) e rótulos (y)
    X = dataframe.drop('targetSentimento', axis=1)  # Considerando que 'rótulos' é a coluna com os rótulos de classificação
    y = dataframe['targetSentimento']

    # Passo 4: Divisão dos dados em conjuntos de treinamento e teste
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Passo 5: Criação e treinamento do modelo Naive Bayes
    model = GaussianNB()
    model.fit(X_train, y_train)

    # Passo 6: Fazer previsões nos dados de teste
    y_pred = model.predict(X_test)

    # Passo 7: Avaliar o modelo
    accuracy = accuracy_score(y_test, y_pred)
    print(f"Acurácia do modelo: {accuracy}")

In [ ]:
# Remover amostras com valores ausentes
df_unido_word.dropna(inplace=True)
df_unido_word

targetSentimento         0         1         2         3         4  \
0             NEUTRAL  1.383377 -2.654717 -1.405555 -3.668075  0.229193   
1             NEUTRAL  1.283763 -3.410798 -2.465253 -3.282195  0.012297   
2            POSITIVE  0.047435 -0.007625  0.173723 -0.445078 -0.058977   
4            POSITIVE  0.106028  0.121924  0.139373 -0.918912  0.004461   
5            NEGATIVE -1.353890 -7.638894 -4.504153 -7.281510  3.319707   
...               ...       ...       ...       ...       ...       ...   
9482         POSITIVE  0.106028  0.121924  0.139373 -0.918912  0.004461   
9483         POSITIVE  1.815789 -2.876974 -1.558797 -3.573577  1.132767   
9484          NEUTRAL  0.836931 -4.082651 -2.630611 -4.756009 -0.629420   
9485          NEUTRAL  0.068602 -8.919511 -2.793904 -7.389237 -0.428109   
9486         NEGATIVE  0.948231 -0.942366 -0.010881 -2.126990  0.949057   

             5         6         7         8  ...        40        41  \
0     0.304061 -1.807550 -0.759527 -4.674443  ...  0.220864 -3.715628   
1    -0.849639 -1.143002 -0.586896 -4.501272  ...  0.604381 -5.294922   
2    -0.182599 -0.079506 -0.220796 -0.265243  ... -0.012755 -0.238091   
4    -0.216214 -0.204989 -0.420545 -0.648111  ... -0.179132 -0.510152   
5    -1.879490 -2.697392  2.624242 -5.669894  ... -0.758122 -5.385909   
...        ...       ...       ...       ...  ...       ...       ...   
9482 -0.216214 -0.204989 -0.420545 -0.648111  ... -0.179132 -0.510152   
9483 -0.111063 -1.510444 -0.775099 -4.272242  ...  0.356593 -4.013429   
9484 -0.144831 -1.697202 -0.302105 -5.067372  ... -0.118279 -6.384274   
9485 -0.655916 -1.413017  0.700860 -5.643440  ... -0.731742 -7.833986   
9486 -0.365847 -0.170032 -1.474218 -2.592782  ...  0.838269 -2.599074   

            42        43        44        45        46        47         48  \
0    -0.508049 -1.681060  0.322339  3.070996 -1.145615 -0.220174  -1.008628   
1    -0.893327 -0.872390  1.264841  2.869610 -0.396956 -0.465566  -2.063515   
2     0.309302 -0.140818 -0.078406 -0.180135  0.013944 -0.043409   0.288535   
4     0.635353 -0.319989 -0.392927  0.184347 -0.225403  0.008890   0.201078   
5    -2.474345  0.160686  5.968213  7.675878 -1.774800 -1.326604 -10.794442   
...        ...       ...       ...       ...       ...       ...        ...   
9482  0.635353 -0.319989 -0.392927  0.184347 -0.225403  0.008890   0.201078   
9483 -0.626634 -1.653492  0.486769  2.573709  0.071672 -0.287472  -0.189655   
9484 -0.833756  0.019192  1.295962  3.993481  0.281847  0.516516  -3.820503   
9485 -0.155092 -0.787291  3.670361  7.178771 -0.241885 -0.687680  -7.578391   
9486  0.021309 -0.552362  0.259679  0.921726 -0.890135  1.305274   0.611135   

            49  
0     4.302267  
1     4.795250  
2     0.232324  
4     0.354131  
5     3.093618  
...        ...  
9482  0.354131  
9483  3.710586  
9484  6.278456  
9485  5.693573  
9486  1.765292  

[8124 rows x 51 columns]

In [ ]:
train_naive_bayes_model_word(df_unido_word)

Acurácia do modelo: 0.36


##### guto

In [ ]:
#verificar quais indices tem um valor nulo
indices_nan = dfWord2Vec[dfWord2Vec[1].isna()].index

dfresultadoModeloWord2Vec = dfWord2Vec.drop(indices_nan)
df = df.drop(indices_nan)

#Traduzir as tuplas que contem negativo para 0 e as que contem neutro e positivo para 1
df['targetSentimento'] = df['targetSentimento'].replace({'POSITIVE': 0, 'NEUTRAL': 0, 'NEGATIVE': 1})

KeyError: ignored

In [ ]:
previsores = dfresultadoModeloWord2Vec
classe = df['targetSentimento']

classe = classe.ravel()

X_train, X_test, y_train, Y_test = train_test_split (previsores, classe, test_size=0.5)

ValueError: ignored

In [ ]:
model = GaussianNB()

In [ ]:
model.fit(X_train, y_train)

In [ ]:
model.score(X_test,Y_test)

### Rede Neural com Word2Vec

In [ ]:
#Importacao Necessárias

import keras
from keras.models import Sequential
from keras.layers import Dense
from sklearn.metrics import confusion_matrix, accuracy_score
from imblearn.over_sampling import SMOTE
from keras.utils import to_categorical

In [ ]:
# Organizar os previsores e classes
previsores = dfresultadoModeloWord2Vec
classe = df['targetSentimento']
classe = classe.drop(indices_nan)


In [ ]:
#Balanciamento das classes ultilizando a biblioteca smote
smote = SMOTE()
X_resampled, y_resampled = smote.fit_resample(previsores, classe)

previsores = pd.DataFrame(X_resampled)
classe = pd.DataFrame(y_resampled)

ValueError: ignored

In [ ]:
#Divisão treinamento e teste
previsores_treinamento, previsores_teste, classe_treinamento, classe_teste = train_test_split(previsores,classe, test_size = 0.25)

In [ ]:
#Criação e treinamento da Rede Neural

classificador = Sequential()

classificador.add(Dense(units = 26, activation = "relu",
                        kernel_initializer = 'random_uniform', input_dim = 50))

classificador.add(Dense(units = 1, activation = 'sigmoid'))


classificador.compile(optimizer = 'adam', loss = 'binary_crossentropy',
                      metrics = ['accuracy'])

classificador.fit(previsores_treinamento, classe_treinamento,
                  batch_size = 1000, epochs = 30)

In [ ]:
# Teste de performance do modelo

previsoes = classificador.predict(previsores_teste)
previsoes = (previsoes> 0.5)


precisao = accuracy_score(classe_teste, previsoes)
matriz = confusion_matrix(classe_teste, previsoes)


resultado = classificador.evaluate(previsores_teste, classe_teste)